server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -239,9 +239,50 @@ private Object fromGenerated(GeneratedReference generated, IndexItem item) {
     }
 
     /**
-     * For DEFAULT expressions or GENERATED columns
+     * For DEFAULT expressions or GENERATED columns.
+     *
+     * <p>Computed values are stored and re-used per-row for:</p>
+     * <ul>
+     *  <li>Sending values of non-deterministic functions to replica</li>
+     *  <li>Computing RETURNING expression, referring to GENERATED or DEFAULT columns.</li>
+     * </ul>
+     *
+     * Cached value must be cleared per-row.
      **/
-    record Synthetic(Reference ref, Input<?> input, ValueIndexer<Object> indexer) {
+    public class Synthetic implements Input<Object> {
+
+        private final Reference ref;
+        private final Input<?> input;
+        private final ValueIndexer<Object> indexer;
+        private Object computedValue;
+
+        public Synthetic(Reference ref,
+                         Input<?> input,
+                         ValueIndexer<Object> indexer) {
+            this.ref = ref;
+            this.input = input;
+            this.indexer = indexer;
+        }
+
+        public Reference ref() {
+            return ref;
+        }
+
+        public ValueIndexer<Object> indexer() {
+            return indexer;
+        }
+
+        @Override
+        public Object value() {
+            if (computedValue == null) {
```

Comment:
I think this needs a dedicated flag to check if it is computed, in case a non-deterministic function can return null or a value.

Reply:
Added flag in https://github.com/crate/crate/pull/14762/commits/fa99342bc6cae62ee9fcff481a63878be5a453fb.

Looks like we currently don't have non-deterministic functions, returning value or null - do you mean we need such flag to be future proof?

Reply:
>  do you mean we need such flag to be future proof in case we add such a function?

Or user defined functions